core/src/main/java/org/elasticsearch/common/limit/LimitingStreamOutput.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.limit;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+/**
+ * A stream wrapper that limits the number of bytes that can be written. The reserved amount is reduced when the stream is closed or reset.
+ */
+public class LimitingStreamOutput extends StreamOutput {
+    private final StreamOutput delegate;
+    private final Limiter limiter;
+    private long bytesWritten;
+
+    public LimitingStreamOutput(StreamOutput delegate, Limiter limiter) {
+        this.delegate = delegate;
+        this.limiter = limiter;
+    }
+
+    @Override
+    public void writeByte(byte b) throws IOException {
+        addByteChecked();
+        delegate.writeByte(b);
+    }
+
+    @Override
+    public void writeBytes(byte[] b, int offset, int length) throws IOException {
+        addBytesChecked(length);
+        delegate.writeBytes(b, offset, length);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        delegate.flush();
+    }
+
+    @Override
+    public void close() throws IOException {
+        addBytesChecked(-bytesWritten);
+        delegate.close();
+    }
+
+    @Override
+    public void reset() throws IOException {
+        addBytesChecked(-bytesWritten);
```

Comment:
As discussed I think both approaches are ok. If we're not reducing the amount in `#reset()` we're a bit more conservative (as we keep the count although there may not be more memory allocated) but when `#close()` is called we'll always free all written bytes, even if `#reset()` has been called in between. So I'll change the implementation and don't reduce the count in `#reset()` anymore.